### PR TITLE
[codegen/go] Support nested collection types

### PR DESCRIPTION
**Pull request by viveklak**
Fixes https://github.com/pulumi/pulumi-azure-native/issues/353
Fixes https://github.com/pulumi/pulumi/issues/6297

With this we are able to cover all constructs required by azure-native go SDK be compiled. I was able to confirm by compiling all the top-level namespaces locally.

*Update* - this PR is ready for review

### CODE REVIEW

**Comment by mikhailshilkov** [comment]
I'm not fluent enough with our Go SDK structure to confidently review this. I'd love others to take a look - it would be great to land the fix soon!

File: pkg/codegen/go/gen.go [hunk 1]
```diff
@@ -507,7 +497,7 @@ func printCommentWithDeprecationMessage(w io.Writer, comment, deprecationMessage
 	}
 }
 
-func genInputInterface(w io.Writer, name string) {
+func genInputInterface(w io.Writer, name, argsName string) {
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go:500]: Looks like the new parameter here is unused?

Comment (viveklak) [reply pkg/codegen/go/gen.go]: Sorry it took a while to return to this. Yes - fixed now. I also fixed a niggling bug with some usage documentation generation which was causing unnecessary changes in some providers.

**Comment by viveklak** [comment]
The PR is rebased and updated/ready for another look. Verified that the only provider with changes is [pulumi-azure-native](https://github.com/pulumi-bot/pulumi-azure-native/compare/875bfa4884..7ab8bfc82e). I also verified that the generated SDK now compiles.

**Review by EvanBoyle — APPROVED** [review]
I appreciate the refactoring that you did in gen.go! Made this change fairly easy to follow along as far as go codegen changes go 😄  and the tests/stable diffs seem to speak for themselves.

**Pull request merged**